### PR TITLE
MINOR: Add upgrade notes for new consumer poll

### PR DESCRIPTION
**Pull request by hachikuji**
Added upgrade notes for the new poll() API. I also added a few small cleanups from #4855.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @vvcephei

**Comment by ijuma** [comment]
I added a couple of comments regarding the documentation here:

https://github.com/apache/kafka/commit/c470ff70d3e829c8b12f6eb6cc812c4162071a1f#diff-e9c1ee46a19a8684d9d8d8a8c77f9005

Generally, it seems like the commit description is very thorough, but the documentation itself is sparse. Are we sure that some of the context in the commit description is not useful for users who have to:

1. Choose between these two methods
2. Understand the implications of migrating to the new method?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Ah, yeah, I forgot about these.

**Review by vvcephei — APPROVED** [review]
Thanks.

**Comment by vvcephei** [comment]
I agree with @ijuma that we could elaborate more in the JavaDoc in `KafkaConsumer`.

On `poll(long)`, maybe we can extend the deprecation notice to say: 
```
@deprecated Since 2.0. Use {@link #poll(Duration)}, which doesn't block indefinitely on metadata. See KIP-266 for more details.
```

On `poll(Duration)`, we can extend the JavaDoc to say:
```
... Performs asynchronous metadata and fetch and awaits responses for {@code timeout} amount of time. 
```

I'm not sure if more detail is necessary here.

**Comment by hachikuji** [comment]
@ijuma @vvcephei Thanks. I added a couple notes on the blocking behavior. Let me know what you think.

**Review by guozhangwang — COMMENTED** [review]
Could we piggy-back the comments I had in https://github.com/apache/kafka/pull/4855#discussion_r191495395 and https://github.com/apache/kafka/pull/4855#discussion_r191496127 in this PR as well?

**Comment by guozhangwang** [comment]
Otherwise, LGTM.

**Comment by hachikuji** [comment]
@guozhangwang Thanks, I modified the field names of `PendingCommittedOffsetRequest` as suggested. About the `remainingTimeAtLeastZero` functions, I'm kind of hoping that we can merge #5087, which merges the behavior of the three functions into the `Timer` class.

**Comment by hachikuji** [comment]
@guozhangwang Hold off on merging this. I will have a few additional changes after I merge  #5014.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:303]: Do you need this currentTime because `ensureCoordinatorReady`, `ensureFreshMetadata` and `ensureActiveGroup` are somewhat long-running?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:303]: Yes, right. I think the approach in #5087 is less irritating.

**Review by vvcephei — APPROVED** [review]
Looks good to me! Thanks for this patch.

**Comment by guozhangwang** [comment]
LGTM! thanks.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1325]: Could we update the KIP wiki: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=75974886#KIP-266:AddTimeoutExceptiontoKafkaConsumer#position()-Consumer#committedandConsumer#commitSync

as well?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1325]: Ack, will do. I've had the editor on that wiki open since this morning, but keep getting distracted.

**Pull request merged**
